### ISSUE TITLE
Tracking: Testing the Entire Chain

### ISSUE THREAD

**Issue by teor2345** [ZcashFoundation/zebra#745]
**Goals**

Make sure that `zebrad`'s sync, verify, and state handling work after every commit.

Check historical edge-case handling in our verifiers, by running the entire chain through them.

These basic tests are easy to automate, and provide reasonable assurance that our implementation works.

# Issues from Epic

- [x] #1021
- [x] #1052
- [x] #1135
- [x] #1392
- [x] #1566
- [x] #2559
- [x] #2560
- [x] #2654
- [x] #208
- [x] #704
- [x] #466
- [x] #1414
- [x] #1302
- [x] #1004
- [x] #919
- [x] #730
- [x] #744
- [x] #610
- [x] #1400
- [ ] #1011
- [ ] #1007
- [ ] #1222
- [ ] #1285
- [ ] #1286
- [ ] #1229

# Original list

#### Implementation
- [x] add a checkpoint_sync config option (#931)
- [x] write acceptance tests for sync, verify, and state (#746)

#### Fix Bugs That Impact Tests
- Tracked in Sync Correctness: #884
- [x] commit blocks according to the state design RFC (#969)
- [x] make the state handle out of order blocks

#### Chain Testing

- [x] Re-enable the CheckpointVerifier tests https://github.com/ZcashFoundation/zebra/issues/1285
  - we might need to use FuturesUnordered to avoid race conditions and deadlocks
  - use Mainnet blocks 11-27 in these tests (added in #802)
- [x] Add mainnet and testnet block test vectors for each network upgrade (#1096)

- [ ] Find existing tests that use block test vectors, and see if they can use `zebra_test::vectors::TEST_BLOCKS` instead

#### Sync Testing
- [x] Test post-Sapling Mainnet blocks (PR #1229)
- [x] Test sync-to-tip on Mainnet
  - as a "canary" test that can only fail https://github.com/ZcashFoundation/zebra/issues/1286 

#### Failure Testing
- [ ] Make sure that the first alpha difficulty validation consensus rules fail on invalid data (#802)

#### Automation
- [x] Integrate quick tests into PR and main branch CI - fast tests
- [x] Deploy slower tests using main branch CD - slow tests

# Lower Priority - Later Releases

#### Design and Scope
- [ ] Network Integration test framework #466
  - [ ] Basic integration testing - scope and design - ticket #1004 RFC #1007
  - [ ] Make zebrad work with a small number of peers #704
  - [ ] Test that zcashd and Zebra can sync from Zebra #1400

#### Fix Bugs That Impact Tests
- Tracked in Sync Correctness: #884
- [x] commit blocks according to the state design RFC (#969)
- [x] make the state handle out of order blocks

#### Chain Testing
- [ ] Re-enable any tests that have been disabled (previous PR #1119)
  - [ ] Re-enable the Block Verifier tests (#1011, depends on #1135)

#### Sync Testing
- Test post-Sapling and sync-to-tip on Testnet
  - [x] Deploy `zcashd` testnet instances to improve testnet reliability (#1222)
    - [x] After `zebrad` can serve blocks (#1078), replace some `zcashd` instances with `zebrad`?
  - [ ] Enable the post-Sapling and sync-to-tip tests by default on Testnet
    - for context, see #1229

#### Failure Testing
- [ ] Make sure that every consensus rule fails on invalid data
- [ ] Make sure that signatures fail on invalid data
- [ ] Make sure that proofs fail on invalid data

**Comment by hdevalence** [comment]
I think our code will be simpler if we don't add any additional modes beyond what we absolutely need.

**Comment by teor2345** [comment]
> I think our code will be simpler if we don't add any additional modes beyond what we absolutely need.

I'd like to test the block verifier against all the blocks in the chain, to make sure we've covered any historical edge cases.

What's the best way to do that?

I could delete some of the hard-coded checkpoints, but that might cause issues with pre-Sapling blocks.

**Comment by teor2345** [comment]
Updated based on #911 and #931.

**Comment by dconnolly** [comment]
#1285

**Comment by teor2345** [comment]
We've moved all of the remaining sub-tasks out of the first alpha release, so this ticket can move to the first stable release.

**Comment by mpguerra** [comment]
Converting any "Epics" that do not have a start and end date into github tracking issues instead.

**Comment by teor2345** [comment]
All these tasks are unlikely to be done any time soon.
